homeassistant/components/motionblinds_ble/sensor.py

```diff
@@ -0,0 +1,221 @@
+            self._attr_icon = "mdi:battery-unknown"
+        elif is_wired:
+            # Motor is wired and does not have a battery
+            self._attr_icon = "mdi:power-plug-outline"
+        elif battery_percentage > 90 and not is_charging:
+            # Full battery icon if battery > 90% and not charging
+            self._attr_icon = "mdi:battery"
+        elif battery_percentage <= 5 and not is_charging:
+            # Empty battery icon with alert if battery <= 5% and not charging
+            self._attr_icon = "mdi:battery-alert-variant-outline"
+        else:
+            battery_icon_prefix = (
+                "mdi:battery-charging" if is_charging else "mdi:battery"
+            )
+            battery_percentage_multiple_ten = ceil(battery_percentage / 10) * 10
+            self._attr_icon = f"{battery_icon_prefix}-{battery_percentage_multiple_ten}"
+        self.async_write_ha_state()
+
+
+class ConnectionSensor(MotionblindsBLESensorEntity):
+    """Representation of a connection sensor entity."""
+
+    async def async_added_to_hass(self) -> None:
+        """Register device callbacks."""
+        await super().async_added_to_hass()
+        self.device.register_connection_callback(self.async_update_connection)
+
+    @callback
+    def async_update_connection(
+        self, connection_type: MotionConnectionType | None
+    ) -> None:
+        """Update the connection sensor value."""
+        self._attr_native_value = connection_type.value if connection_type else None
+        self.async_write_ha_state()
+
+
+class CalibrationSensor(MotionblindsBLESensorEntity):
+    """Representation of a calibration sensor entity."""
+
+    async def async_added_to_hass(self) -> None:
+        """Register device callbacks."""
+        await super().async_added_to_hass()
+        self.device.register_calibration_callback(self.async_update_calibration)
+
+    @callback
+    def async_update_calibration(
+        self, calibration_type: MotionCalibrationType | None
+    ) -> None:
+        """Update the calibration sensor value."""
+        self._attr_native_value = (
+            calibration_type.value if calibration_type is not None else None
+        )
+        self.async_write_ha_state()
+
+
+class SignalStrengthSensor(MotionblindsBLESensorEntity):
+    """Representation of a signal strength sensor entity."""
+
+    async def async_added_to_hass(self) -> None:
+        """Register device callbacks and update signal strength."""
+        await super().async_added_to_hass()
+        self.device.register_signal_strength_callback(self.async_update_signal_strength)
+        self.async_update_signal_strength(self.device.rssi)
+
+    @callback
+    def async_update_signal_strength(self, signal_strength: int | None) -> None:
+        """Update the signal strength sensor value."""
+        self._attr_native_value = (
+            str(signal_strength) if signal_strength is not None else None
+        )
+        self.async_write_ha_state()
```

Comment:
I think it would be nice if we tried to make one generic class in which we can change the entity type by injecting the entity description. That's the optimal way of using the entity description.

Reply:
What would that look like? I already use a `MotionblindsBLESensorEntity` class that sets the entity_description

Reply:
I assume this is not really necessary since I am already passing in the entity description? I don't know how I could make it much more compact.

Reply:
I am leaning towards or just trying to make one sensor as generic as possible. But maybe it makes sense to get rid of the entity descriptions in the first place. The idea of entity descriptions is that you can inject an entity description into an entity class and it will start behaving like that entity description. But the downside with the current approach, you can't inject the Battery entity description in the signal strength one, because that doesn't make sense, so the value that having entity descriptions have here, is very very little

Reply:
What would you suggest doing? Would the other option be to set everything through `self` in the constructor? If so, I think it's still nice to have all that sensor description information at the top of the file where it can be easily found and changed.

I don't know if I can make it much more compact without making it more complicated. For sure there would be a way to reference the `register` function on the device and then a method that is used to update the value of the sensor, but that would probably make it quite a bit harder to understand.

Reply:
Definetly, you can add stuff as shorthand attribute:

```py
class BatterySensor(MotionblindsBLESensorEntity):
    """Representation of a battery sensor entity."""

    _attr_device_class = SensorDeviceClass.BATTERY
    _attr_translation_key = "battery"
    _attr_state_class=SensorStateClass.MEASUREMENT
    _attr_entity_category = EntityCategory.DIAGNOSTIC

    async def async_added_to_hass(self) -> None:
        """Register device callbacks."""
        await super().async_added_to_hass()
        self.device.register_battery_callback(self.async_update_battery)
```

Imo, the power of the entity descriptions is to have all the things that define an entity in one place, this way you still have this.

I can also mark this as second opinion wanted if you want to have a second pair of eyes on this matter

Reply:
If putting all of the attributes that are in the entity description in the classes themselves solves this, then I don't mind doing that

Reply:
But also let me know what you think, I think this still makes it easy to update a certain value, and it doesn't look confusing as like "oh, the entity definition is here, but where is the logic??"

Reply:
I agree, may be better to just put it in the class itself. Don't I need to define `native_unit_of_measurement` and `key` using self._attr_ as you did not include it in your example?

Reply:
By default everything except for `key` means something special, although, you can add an `_attr_key` yourself to use in the unique_id

Reply:
Actually, currently the `MotionblindsBLEEntity` class takes in the entity_description. Therefore, I prefer to keep it consistent with the other button and select entities and pass in an entity_description instead of defining it on the class if that is possible.

```python
class MotionblindsBLEEntity(Entity):
    """Base class for Motionblinds BLE entities."""

    _attr_has_entity_name = True
    _attr_should_poll = False

    device: MotionDevice
    entry: ConfigEntry

    def __init__(
        self,
        device: MotionDevice,
        entry: ConfigEntry,
        entity_description: EntityDescription,
        unique_id_suffix: str | None = None,
    ) -> None:
```

Reply:
Maybe checkout Philips hue, they have a different way of using entity descriptions, maybe that would work here as well

Reply:
The Philips hue integration sets `entity_description` in the class body. If I were to do that I would still have to change the `MotionblindsBLESensorEntity` class and then it would be used differently for buttons, select and sensors.

What I could do is add this constructor to all sensors, then you would still have the entity description definition in the class itself, but in the constructor instead of the body:
```python
    def __init__(
        self,
        device: MotionDevice,
        entry: ConfigEntry,
    ) -> None:
        """Initialize the sensor entity."""
        entity_description = EntityDescription(...)
        super().__init__(
            device, entry, entity_description, unique_id_suffix=entity_description.key
        )
```

Reply:
These sensor can use a base class together with extending the entity description

The code is untested as I don't have any device:
```python
"""Sensor entities for the Motionblinds BLE integration."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
import logging
from math import ceil
from typing import Generic, TypeVar

from motionblindsble.const import (
    MotionBlindType,
    MotionCalibrationType,
    MotionConnectionType,
)
from motionblindsble.device import MotionDevice

from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
    SensorStateClass,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    PERCENTAGE,
    SIGNAL_STRENGTH_DECIBELS_MILLIWATT,
    EntityCategory,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import StateType

from .const import (
    ATTR_BATTERY,
    ATTR_CALIBRATION,
    ATTR_CONNECTION,
    ATTR_SIGNAL_STRENGTH,
    CONF_MAC_CODE,
    DOMAIN,
)
from .entity import MotionblindsBLEEntity

_LOGGER = logging.getLogger(__name__)

PARALLEL_UPDATES = 0

_T = TypeVar("_T")


@dataclass(frozen=True, kw_only=True)
class MotionblindsBLESensorEntityDescription(SensorEntityDescription, Generic[_T]):
    """Entity description of a sensor entity with initial_value attribute."""

    initial_value: str | None = None
    register_callback_func: Callable[[MotionDevice], Callable[[_T | None], None]]
    value_func: Callable[[_T | None], StateType]
    is_supported: Callable[[MotionDevice], bool] = lambda device: True


SENSORS: tuple[MotionblindsBLESensorEntityDescription, ...] = (
    MotionblindsBLESensorEntityDescription[MotionConnectionType](
        key=ATTR_CONNECTION,
        translation_key=ATTR_CONNECTION,
        device_class=SensorDeviceClass.ENUM,
        entity_category=EntityCategory.DIAGNOSTIC,
        options=["connected", "connecting", "disconnected", "disconnecting"],
        initial_value=MotionConnectionType.DISCONNECTED.value,
        register_callback_func=lambda device: device.register_connection_callback,
        value_func=lambda value: value.value if value else None,
    ),
    MotionblindsBLESensorEntityDescription[MotionCalibrationType](
        key=ATTR_CALIBRATION,
        translation_key=ATTR_CALIBRATION,
        device_class=SensorDeviceClass.ENUM,
        entity_category=EntityCategory.DIAGNOSTIC,
        options=["calibrated", "uncalibrated", "calibrating"],
        register_callback_func=lambda device: device.register_calibration_callback,
        value_func=lambda value: value.value if value else None,
        is_supported=lambda device: device.blind_type
        in {MotionBlindType.CURTAIN, MotionBlindType.VERTICAL},
    ),
    MotionblindsBLESensorEntityDescription[int](
        key=ATTR_SIGNAL_STRENGTH,
        device_class=SensorDeviceClass.SIGNAL_STRENGTH,
        entity_category=EntityCategory.DIAGNOSTIC,
        native_unit_of_measurement=SIGNAL_STRENGTH_DECIBELS_MILLIWATT,
        register_callback_func=lambda device: device.register_signal_strength_callback,
        value_func=lambda value: value,
    ),
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up sensor entities based on a config entry."""

    device: MotionDevice = hass.data[DOMAIN][entry.entry_id]

    entities: list[SensorEntity] = [
        MotionblindsBLESensorEntity(device, entry, description)
        for description in SENSORS
        if description.is_supported(device)
    ]
    entities.append(BatterySensor(device, entry))
    async_add_entities(entities)


class MotionblindsBLESensorEntity(MotionblindsBLEEntity, SensorEntity, Generic[_T]):
    """Representation of a sensor entity."""

    entity_description: MotionblindsBLESensorEntityDescription[_T]

    def __init__(
        self,
        device: MotionDevice,
        entry: ConfigEntry,
        entity_description: MotionblindsBLESensorEntityDescription[_T],
    ) -> None:
        """Initialize the sensor entity."""
        super().__init__(
            device, entry, entity_description, unique_id_suffix=entity_description.key
        )
        self._attr_native_value = entity_description.initial_value

    async def async_added_to_hass(self) -> None:
        """Log sensor entity information."""
        _LOGGER.debug(
            "(%s) Setting up %s sensor entity",
            self.entry.data[CONF_MAC_CODE],
            self.entity_description.key.replace("_", " "),
        )

        def async_callback(value: _T | None) -> None:
            """Update the sensor value."""
            self._attr_native_value = self.entity_description.value_func(value)
            self.async_write_ha_state()

        self.entity_description.register_callback_func(self.device)(async_callback)


class BatterySensor(MotionblindsBLEEntity, SensorEntity):
    """Representation of a battery sensor entity."""

    def __init__(
        self,
        device: MotionDevice,
        entry: ConfigEntry,
    ) -> None:
        """Initialize the sensor entity."""
        entity_description = SensorEntityDescription(
            key=ATTR_BATTERY,
            translation_key=ATTR_BATTERY,
            native_unit_of_measurement=PERCENTAGE,
            device_class=SensorDeviceClass.BATTERY,
            state_class=SensorStateClass.MEASUREMENT,
            entity_category=EntityCategory.DIAGNOSTIC,
        )
        super().__init__(device, entry, entity_description)

    async def async_added_to_hass(self) -> None:
        """Register device callbacks."""
        await super().async_added_to_hass()
        self.device.register_battery_callback(self.async_update_battery)

    @callback
    def async_update_battery(
        self,
        battery_percentage: int | None,
        is_charging: bool | None,
        is_wired: bool | None,
    ) -> None:
        """Update the battery sensor value and icon."""
        self._attr_native_value = battery_percentage
        )
        if battery_percentage is None:
            # Battery percentage is unknown
            self._attr_icon = "mdi:battery-unknown"
        elif is_wired:
            # Motor is wired and does not have a battery
            self._attr_icon = "mdi:power-plug-outline"
        elif battery_percentage > 90 and not is_charging:
            # Full battery icon if battery > 90% and not charging
            self._attr_icon = "mdi:battery"
        elif battery_percentage <= 5 and not is_charging:
            # Empty battery icon with alert if battery <= 5% and not charging
            self._attr_icon = "mdi:battery-alert-variant-outline"
        else:
            battery_icon_prefix = (
                "mdi:battery-charging" if is_charging else "mdi:battery"
            )
            battery_percentage_multiple_ten = ceil(battery_percentage / 10) * 10
            self._attr_icon = f"{battery_icon_prefix}-{battery_percentage_multiple_ten}"
        self.async_write_ha_state()
```

Reply:
Interesting solution @edenhaus, I applied it. I don't think the code has become much easier to read or much smaller, but I don't mind doing it like this